### PR TITLE
Add SQL over HTTP endpoint in proxy

### PR DESCRIPTION
**Pull request by kelvich**
That is a next attempt to get json ednpoint in proxy. In the current approach:
* we use modified rust-postgres driver to
  * avoid checking input types as driver not always catching up with postgres typing rules, e.g. https://github.com/sfackler/rust-postgres/issues/1030
  * use text protocol to support values that do not have binary representation, like aclitem in pg_class
  * avoids extra roundtrip to describe and check argument types
* since we use text protocol manual array parsing code was introduced
* current type mappings are following:
  * pg int2/int4/float4/float8 -> json number (NB: NaN and Inf are left as text, whether JSON.stringify converts to nulls)
  * bool, null, text -> bool, null, string
  * array -> array
  * json -> json
* some amount of unit integration tests
* check SNI

open item:
- [x] change response format to pass oid

possible follow ups:
- [ ] check that http error codes are okay
- [ ] streaming mode -- avoid materializing all rows and stream as a response instead (To avoid OOM on a big query)
- [ ] look up what startup packet settings node-postgres sends (timezone, date format, etc) and allow them in headers too
- [ ] separate auth params support (as well as connection string)

### CODE REVIEW

File: proxy/src/http/sql_over_http.rs [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+        }
+    }
+}
+
+fn get_conn_info(
+    headers: &HeaderMap,
+    sni_hostname: Option<String>,
+) -> Result<(String, String, String, String), anyhow::Error> {
+    let connection_string = headers
+        .get("X-Neon-ConnectionString")
```
Comment (vadim2404) [inline proxy/src/http/sql_over_http.rs:80]: `X-Neon-Connection-String` maybe sounds better? Two different types of naming are mixed.

File: test_runner/fixtures/neon_fixtures.py [hunk 2]
```diff
@@ -2042,15 +2042,40 @@ def __init__(
...
         auth_backend: NeonProxy.AuthBackend,
         metric_collection_endpoint: Optional[str] = None,
         metric_collection_interval: Optional[str] = None,
     ):
         host = "127.0.0.1"
-        super().__init__(dsn=auth_backend.default_conn_url, host=host, port=proxy_port)
+        domain = "proxy.localtest.me" # resolves to 127.0.0.1
+        super().__init__(dsn=auth_backend.default_conn_url, host=domain, port=proxy_port)
 
+        subprocess.run(
```
Comment (vadim2404) [inline test_runner/fixtures/neon_fixtures.py:2054]: check the output code

and running processing in `__init__` is not a common practice.

File: proxy/src/http/sql_over_http.rs [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+        .collect::<Result<Map<String, Value>, anyhow::Error>>()?;
+
+    Ok(Value::Object(res))
+}
+
+//
+// Convert postgres text-encoded value to JSON value
+//
+pub fn pg_text_to_json(pg_value: Option<&str>, pg_type: &Type) -> Result<Value, anyhow::Error> {
+    if let Some(val) = pg_value {
```
Comment (vadim2404) [inline proxy/src/http/sql_over_http.rs:242]: I thought, maybe it'll be easier to wrap incoming queries with `json_agg(row)` and not to parse all these data :D

File: test_runner/regress/test_proxy.py [hunk 4]
```diff
@@ -140,26 +161,45 @@ def test_forward_params_to_client(static_proxy: NeonProxy, option_name: str):
...
+        response = requests.post(
+            f"https://{static_proxy.domain}:{static_proxy.external_http_port}/sql",
+            data=json.dumps({"query": sql, "params": params}),
+            headers={"Content-Type": "application/sql", "X-Neon-ConnectionString": connstr},
+            verify= str(static_proxy.test_output_dir / "proxy.crt"),
+        )
+        assert response.status_code == 200
+        return response.json()
+
+    assert q('select 42 as answer') == [{'answer': 42}]
```
Comment (vadim2404) [inline test_runner/regress/test_proxy.py:200]: what's about `update`/`delete` queries?

File: proxy/src/http/sql_over_http.rs (hunk 1 above)
Comment (bayandin) [reply proxy/src/http/sql_over_http.rs:80]: Maybe we should drop `X-` prefix as well, as it's not recommended anymore?

https://datatracker.ietf.org/doc/html/rfc6648

**Comment by kelvich** [comment]
fyi: there is also https://github.com/neondatabase/rust-postgres/pull/14

File: test_runner/regress/test_proxy.py (hunk 4 above)
Comment (kelvich) [reply test_runner/regress/test_proxy.py]: Added tests for DDL and DML.

Unexpectedly, rust-postgres hadn't returned command tags, so any DML was quite blind. Changed driver one more time https://github.com/neondatabase/rust-postgres/pull/14/commits/7d3933c362c5142a0668bc205584485f17a61976

Seems that @skyzh stepped on the same thing already: https://github.com/sfackler/rust-postgres/pull/985/files (btw there is a bug -- "INSERT" unlike other tags returns two numbers where the first one is always zero and the second one is actual number of rows, https://www.postgresql.org/docs/current/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-COMMANDCOMPLETE)

In my case I needed command tag string too so did that in a way that is straightforward, but not compatible with the method that is currently in the latest version. A bit more work need to proper upstream that patch.

File: vendor/postgres-v14
[the remark this answers is not in the document]
Comment (kelvich) [reply vendor/postgres-v14:1]: that was accidental, fixed

File: proxy/src/http/sql_over_http.rs (hunk 3 above)
Comment (kelvich) [reply proxy/src/http/sql_over_http.rs:287]: Parsing, manipulating and serializing SQL seems to be harder : ) However there is https://github.com/pganalyze/pg_query.rs so doable

File: test_runner/fixtures/neon_fixtures.py (hunk 2 above)
Comment (kelvich) [reply test_runner/fixtures/neon_fixtures.py]: moved to start and checked ret code

File: proxy/src/http/sql_over_http.rs (hunk 1 above)
Comment (kelvich) [reply proxy/src/http/sql_over_http.rs]: done

File: test_runner/regress/test_proxy.py (hunk 4 above)
Comment (vadim2404) [reply test_runner/regress/test_proxy.py]: what about DML queries with the `RETURNING` option?

Comment (kelvich) [reply test_runner/regress/test_proxy.py]: > what about DML queries with the RETURNING option?

no surprises, works as expected, added to tests

File: proxy/src/http/websocket.rs [hunk 5]
```diff
@@ -159,6 +166,7 @@ async fn ws_handler(
     config: &'static ProxyConfig,
     cancel_map: Arc<CancelMap>,
     session_id: uuid::Uuid,
+    sni_hostname: Option<String>,
```
Comment (petuhovskiy) [inline proxy/src/http/websocket.rs:169]: Is it different from `host` 2 lines below? Why sni_hostname is better for SQL-over-HTTP queries?

File: proxy/src/http/sql_over_http.rs [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+    //
+    let conn_string = &format!(
+        "host={} port={} user={} password={} dbname={}",
+        host,
+        conf.get_ports().first().expect("no port"),
+        username,
+        password,
+        dbname
+    );
+    let (client, connection) = tokio_postgres::connect(conn_string, tokio_postgres::NoTls).await?;
```
Comment (hlinnaka) [inline proxy/src/http/sql_over_http.rs:201]: Escaping!

File: proxy/src/http/sql_over_http.rs [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+    //
+    let row_stream = client.query_raw_txt(query, query_params).await?;
+
+    // manually drain the stream into a vector to leave row_stream hanging
+    // around to get a command tag
+    pin_mut!(row_stream);
+    let mut rows: Vec<tokio_postgres::Row> = Vec::new();
+    while let Some(row) = row_stream.next().await {
+        rows.push(row.unwrap());
+    }
```
Comment (hlinnaka) [inline proxy/src/http/sql_over_http.rs:219]: The result set can be arbitrarily large, leading to an out-of-memory panic.

I think the array and json handling code have similar issues.

File: proxy/src/http/sql_over_http.rs [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+        tokio_postgres::config::Host::Tcp(host) => host,
+        tokio_postgres::config::Host::Unix(_) => {
+            return Err(anyhow::anyhow!("unix socket is not supported"));
+        }
+    };
+
+    //
+    // Read the query and query params from the request body
+    //
+    let body = hyper::body::to_bytes(request.into_body()).await?;
```
Comment (petuhovskiy) [inline proxy/src/http/sql_over_http.rs:186]: We need to limit the request length to guard against malicious clients.

**Comment by hlinnaka** [comment]
Some documentation on the query and response JSON objects would be nice

File: proxy/src/http/sql_over_http.rs [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+
+    //
+    // Now execute the query and return the result
+    //
+    let row_stream = client.query_raw_txt(query, query_params).await?;
+
+    // manually drain the stream into a vector to leave row_stream hanging
+    // around to get a command tag
+    pin_mut!(row_stream);
+    let mut rows: Vec<tokio_postgres::Row> = Vec::new();
```
Comment (petuhovskiy) [inline proxy/src/http/sql_over_http.rs:216]: Would also be good to limit the bytes used to store rows in the vector.

File: proxy/src/http/websocket.rs [hunk 10]
```diff
@@ -181,8 +189,41 @@ async fn ws_handler(
 
         // Return the response so the spawned future can continue.
         Ok(response)
+    // TODO: that deserves a refactor as now this function also handles http json client besides websockets.
+    // Right now I don't want to blow up sql-over-http patch with file renames and do that as a follow up instead.
+    } else if request.uri().path() == "/sql" && request.method() == Method::POST {
+        let result = sql_over_http::handle(config, request, sni_hostname).await;
```
Comment (petuhovskiy) [inline proxy/src/http/websocket.rs:195]: We can also add timeout here, to limit extra large queries/responses the same way as it's done in the console.

File: proxy/src/http/websocket.rs (hunk 5 above)
Comment (kelvich) [reply proxy/src/http/websocket.rs:169]: Yes, and SNI host is the only one that is checked against our certs, so we will reject connection with a forged SNI host.

File: proxy/src/http/websocket.rs [hunk 11]
```diff
@@ -181,8 +189,41 @@ async fn ws_handler(
...
+            }
+        };
+        json_response(status_code, json).map(|mut r| {
+            r.headers_mut().insert(
+                "Access-Control-Allow-Origin",
+                hyper::http::HeaderValue::from_static("*"),
+            );
+            r
+        })
+    } else if request.uri().path() == "/sleep" {
```
[the remark this answers is not in the document]
Comment (kelvich) [reply proxy/src/http/websocket.rs:221]: We also can do pg_sleep now, deleting

File: proxy/src/http/sql_over_http.rs (hunk 6 above)
Comment (kelvich) [reply proxy/src/http/sql_over_http.rs:201]: 🤦‍♀️ Started a notion page to collect that examples https://www.notion.so/neondatabase/Potential-vulnerabilities-caught-in-PRs-9f16af6bf06445d19804106d8eccf8ef

File: proxy/src/http/sql_over_http.rs (hunk 7 above)
Comment (kelvich) [reply proxy/src/http/sql_over_http.rs:219]: Yes, streaming mode marked as one of the follow ups -- ideally all of that could work like a pipeline. But no easy to implement. I'm thinking of counting bytes and throwing an error if we exceed 1MB.

**Comment by kelvich** [comment]
> Some documentation on the query and response JSON objects would be nice

pushed

**Comment by hlinnaka** [comment]
I googled around for similar HTTP APIs:

https://docs.snowflake.com/en/developer-guide/sql-api/submitting-requests
https://docs.singlestore.com/db/v8.0/en/reference/data-api/data-api-endpoint-reference.html
https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-rest-overview.html
https://druid.apache.org/docs/latest/querying/sql-api.html
https://docs.starburst.io/latest/develop/client-protocol.html
https://docs.dremio.com/software/api/sql/

I don't think any one of them is popular enough that there would be much value in copying it as is. But maybe look at them for inspiration, and see if they have something important that we're missing here.

**Comment by petuhovskiy** [comment]
There's also planetscale API for the fetch mysql driver. It doesn't have a public specification, but the driver code has some typed structs: https://github.com/planetscale/database-js/blob/main/src/index.ts

File: proxy/src/http/websocket.rs (hunk 10 above)
Comment (kelvich) [reply proxy/src/http/websocket.rs]: done

File: proxy/src/http/sql_over_http.rs (hunk 9 above)
Comment (kelvich) [reply proxy/src/http/sql_over_http.rs:219]: Added 1MB db response limit. One more change to the rust-postgres to get row sizes.

**Comment by kelvich** [comment]
> Some documentation on the query and response JSON objects would be nice

Added

File: proxy/src/http/sql_over_http.rs (hunk 7 above)
Comment (kelvich) [reply proxy/src/http/sql_over_http.rs]: Added size limit here.

**Comment by kelvich** [comment]
So looks like we still have two places that may eat a lot of memory:
* `hyper::body::to_bytes(request.into_body())` is the only place that can grow indefinitely
* single row could also be quite large, which could be problematic. In theory we can count TCP stream bytes over the socket and throw an error
* any other such places?

**Comment by petuhovskiy** [comment]
Added more limits for request and response sizes.

HTTP request body length is checked with `size_hint`. It seems to be using `Content-Length` in most cases, but maybe hyper has some additional hacks.

Postgres row size is limited with a patch in rust-postgres, which sets a limit for all postgres->proxy backend messages.

File: proxy/src/http/sql_over_http.rs (hunk 8 above)
Comment (petuhovskiy) [reply proxy/src/http/sql_over_http.rs:201]: +1 for using WAF, but for now fixed in https://github.com/neondatabase/neon/pull/4272/commits/aab570ea951b6f38e7cd231b9871ac81900f11b6

File: proxy/src/http/sql_over_http.rs [hunk 12]
```diff
@@ -0,0 +1,603 @@
...
+
+    Ok(Value::Object(res))
+}
+
+//
+// Convert postgres text-encoded value to JSON value
+//
+pub fn pg_text_to_json(pg_value: Option<&str>, pg_type: &Type) -> Result<Value, anyhow::Error> {
+    if let Some(val) = pg_value {
+        if val == "NULL" {
```
Comment (petuhovskiy) [inline proxy/src/http/sql_over_http.rs:311]:
```suggestion
        if val == "NULL" && *pg_type != Type::TEXT {
```

To fix `SELECT "NULL"`

Comment (petuhovskiy) [reply proxy/src/http/sql_over_http.rs:311]: Hm, then it doesn't work for `SELECT NULL`...

Comment (petuhovskiy) [reply proxy/src/http/sql_over_http.rs:311]: Should be ok for now.

**Review by petuhovskiy — APPROVED** [review]

**Pull request merged**